### PR TITLE
Add tag exclusion functionality in v1.1

### PR DESCRIPTION
**Pull request by torivara**
Added functionality for allowing excluded VMs based on tags.

Also fixed a few linter errors where null comparisons were switched around. $null should always be on the left side of comparisons.

### CODE REVIEW

File: Runbooks/startstop/1.1/Bootstrap_Main.ps1 [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+#>
+
+
+# ------------------Execution Entry point ---------------------
+
+Write-Output 'Bootstrap main script execution started...'
+
+#---------Inputs variables for NewRunAsAccountCertKeyVault.ps1 child bootstrap script--------------
+$automationAccountName = Get-AutomationVariable -Name 'Internal_AutomationAccountName'
+$SubscriptionId = Get-AutomationVariable -Name 'Internal_AzureSubscriptionId'
```
Comment (Rewdie) [inline Runbooks/startstop/1.1/Bootstrap_Main.ps1:20]: Never used

File: Runbooks/startstop/1.1/ScheduledStartStop_Base_Classic.ps1 [hunk 2]
```diff
@@ -0,0 +1,161 @@
+﻿<#
+.SYNOPSIS
+	This runbook used to perform action start or stop in classic VM group by Cloud Services
+.DESCRIPTION
+	This runbook used to perform action start or stop in classic VM group by Cloud Services
+	This runbook requires the Azure Automation Run-As (Service Principle) account, which must be added when creating the Azure Automation account.
+ .EXAMPLE
```
Comment (Rewdie) [inline Runbooks/startstop/1.1/ScheduledStartStop_Base_Classic.ps1:7]: Extra space at start of line

File: Runbooks/startstop/1.1/ScheduledStartStop_Parent.ps1 [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+	 Wrapper script for get all the VM's in all RG's or subscription level and then call the Start or Stop runbook
+.DESCRIPTION
+	 This runbook is intended to start/stop VMs (both classic and ARM based VMs) that resides in a given list of Azure resource group(s).If the resource group list is empty, then the script gets all the VMs in the current subscription.
+	 Upon completion of the runbook, an option to email results of the started VM can be sent via SendGrid account.
+
+	 This runbook requires the Azure Automation Run-As (Service Principle) account, which must be added when creating the Azure Automation account.
+.EXAMPLE
+	.\ScheduledStartStop_Parent.ps1 -Action "Value1" -WhatIf "False"
+
+.PARAMETER
```
Comment (Rewdie) [inline Runbooks/startstop/1.1/ScheduledStartStop_Parent.ps1:12]: Extra space at start of line

File: Runbooks/startstop/1.1/SequencedStartStop_Parent.ps1 [hunk 4]
```diff
@@ -0,0 +1,744 @@
+﻿<#
+.SYNOPSIS
+	This runbook used to perform sequenced start or stop Azure RM.
+.DESCRIPTION
+	This runbook used to perform sequenced start or stop Azure RM.
+	Create a tag called “sequencestart” on each VM that you want to sequence start activity for.Create a tag called “sequencestop” on each VM that you want to sequence stop activity for. The value of the tag should be an integer (1,2,3) that corresponds to the order you want to start\stop. For both action, the order goes ascending (1,2,3) . WhatIf behaves the same as in other runbooks.
+	Upon completion of the runbook, an option to email results of the started VM can be sent via Action Groups and alerts.
+
+	This runbook requires the Azure Automation Run-As (Service Principle) account, which must be added when creating the Azure Automation account.
+ .PARAMETER
```
Comment (Rewdie) [inline Runbooks/startstop/1.1/SequencedStartStop_Parent.ps1:10]: Extra space at start

**Review by Rewdie — APPROVED** [review]
The v1.0 changes seem to contains some residue of some merge conflicts, and the Autostop_createLaert_Child script looks a bit off.
The v1.1 looks good, can only spot a very minor formatting errors

File: Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1 [hunk 5]
```diff
@@ -136,186 +127,214 @@ try {
...
-                            -MetricName $metricName `
-                            -Operator $condition `
-                            -Threshold $threshold `
-                            -WindowSize $timeWindow `
-                            -TimeAggregationOperator $timeAggregationOperator `
-                            -Action $actionWebhook `
-                            -Description $description -DisableRule
-
-                        Write-Output "Alert ($($Alert.Name)) Disabled for VM $($VMObject.Name)"
+            <<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (torivara) [reply Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1]: Fixed.

File: Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1 [hunk 6]
```diff
@@ -136,186 +127,214 @@ try {
...
+            if ($null -ne $ExVMAlerts) {
+                =======
+                if ($ExVMAlerts -ne $null) {
+                    >>>>>>> 1860b9ebcd4584a0d65d53b8d8a92135eb87b37e
```
[the remark this answers is not in the document]
Comment (torivara) [reply Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1]: Fixed.

File: Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1 [hunk 7]
```diff
@@ -136,186 +127,214 @@ try {
...
-                            -WindowSize $timeWindow `
-                            -TimeAggregationOperator $timeAggregationOperator `
-                            -Action $actionWebhook `
-                            -Description $description -DisableRule
-
-                        Write-Output "Alert ($($Alert.Name)) Disabled for VM $($VMObject.Name)"
+            <<<<<<< HEAD
+            if ($null -ne $ExVMAlerts) {
+                =======
+                if ($ExVMAlerts -ne $null) {
```
[the remark this answers is not in the document]
Comment (torivara) [reply Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1]: Fixed.

File: Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1 [hunk 8]
```diff
@@ -136,186 +127,214 @@ try {
...
+            } elseif ($AlertAction -eq 'Create') {
+                #Getting ResourcegroupName and Location based on VM
+
+                #if (($VMState -eq 'PowerState/running') -or ($VMState -eq 'ReadyRole'))
+                #{
+                Write-Output 'Creating alerts...'
+                $VMAlerts = Get-AzAlertRule -ResourceGroupName $ResourceGroupName -DetailedOutput -ErrorAction SilentlyContinue
+
+                #Check if alerts exists and take action
+                <<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (torivara) [reply Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1]: Fixed.

File: Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1 [hunk 9]
```diff
@@ -136,186 +127,214 @@ try {
...
+                if ($null -ne $VMAlerts) {
+                    =======
+                    if ($VMAlerts -ne $null) {
```
[the remark this answers is not in the document]
Comment (torivara) [reply Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1]: Fixed.

File: Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1 [hunk 10]
```diff
@@ -136,186 +127,214 @@ try {
...
+                        >>>>>>> 1860b9ebcd4584a0d65d53b8d8a92135eb87b37e
```
[the remark this answers is not in the document]
Comment (torivara) [reply Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1]: Fixed.

File: Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1 [hunk 11]
```diff
@@ -136,186 +127,214 @@ try {
...
+                                Write-Output "Previous alert ($($Alert.Name)) found and deleting now..."
+                                #Remove the old alert
+                                Remove-AzAlertRule -Name $Alert.Name -ResourceGroupName $ResourceGroupName
+
+                                #Wait for few seconds to make sure it processed
+                                Do {
+                                    #Start-Sleep 10
+                                    $GetAlert = Get-AzAlertRule -ResourceGroupName $ResourceGroupName -Name $Alert.Name -DetailedOutput -ErrorAction SilentlyContinue
+                                }
+                                <<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (torivara) [reply Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1]: Fixed.

File: Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1 [hunk 12]
```diff
@@ -136,186 +127,214 @@ try {
...
+                        $VMAlerts = Get-AzMetricAlertRuleV2 -ResourceGroupName $ResourceGroupName -ErrorAction SilentlyContinue
 
-                        Write-Output 'Generating a new alert with unique name...'
-                        #Now generate new unique alert name
-                        $NewAlertName = Generate-AlertName -OldAlertName $Alert.Name -VMName $VMObject.Name
-                    }
-                }
+                        #Check if alerts exists and take action
+                        <<<<<<< HEAD
+                        if ($null -ne $VMAlerts) {
```
[the remark this answers is not in the document]
Comment (torivara) [reply Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1]: Fixed.

File: Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1 [hunk 13]
```diff
@@ -136,186 +127,214 @@ try {
...
+                            $actionGroupShortName = 'AutoStopAG'
 
-            #2 Create the action groupId
-            $actionGroupId = New-AzActionGroup -ActionGroupId $actionGroup.Id
+                            $actionGroup = Get-AzActionGroup -Name $actionGroupName -ResourceGroupName $aroResourceGroupName -ErrorAction SilentlyContinue
 
-            #3 Create the condition/criteria
-            $criteria = New-AzMetricAlertRuleV2Criteria -MetricName $metricName -MetricNamespace 'Microsoft.Compute/virtualMachines' -TimeAggregation $timeAggregationOperator -Operator $condition -Threshold $threshold
+                            <<<<<<< HEAD
+                            if ($null -eq $actionGroup) {
```
[the remark this answers is not in the document]
Comment (torivara) [reply Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1]: Fixed.

File: Runbooks/startstop/1.1/AutoStop_CreateAlert_Child.ps1 [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+                            -Location $Alert.Location `
+                            -ResourceGroupName $ResourceGroupName `
+                            -TargetResourceId $resourceId `
+                            -MetricName $metricName `
+                            -Operator $condition `
+                            -Threshold $threshold `
+                            -WindowSize $timeWindow `
+                            -TimeAggregationOperator $timeAggregationOperator `
+                            -Action $actionWebhook `
+                            -Description $description -DisableRule
```
[the remark this answers is not in the document]
Comment (torivara) [reply Runbooks/startstop/1.1/AutoStop_CreateAlert_Child.ps1:147]: Fixed.

File: Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1 [hunk 15]
```diff
@@ -136,186 +127,214 @@ try {
...
+                                    -TargetResourceId $resourceId -Condition $criteria -ActionGroup $actionGroupId -Severity $severity -Description $description
+
+
+                                Write-Output "Alert Created for VM $($VMObject.Name.Trim())"
+                            }
+                        }
+                    } catch {
+                        Write-Output 'Error Occurred'
+                        Write-Output $_.Exception
+                    }
```
[the remark this answers is not in the document]
Comment (torivara) [reply Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1]: Not sure what you mean here. No errors on indentation or loops in my editor. Please elaborate, @Rewdie

File: Runbooks/startstop/1.1/Bootstrap_Main.ps1 (hunk 1 above)
Comment (torivara) [reply Runbooks/startstop/1.1/Bootstrap_Main.ps1:20]: Not important. I will leave it as is for now.

File: Runbooks/startstop/1.1/ScheduledStartStop_Base_Classic.ps1 (hunk 2 above)
Comment (torivara) [reply Runbooks/startstop/1.1/ScheduledStartStop_Base_Classic.ps1:7]: Fixed something. It is not important anyway.

File: Runbooks/startstop/1.1/SequencedStartStop_Parent.ps1 (hunk 4 above)
Comment (torivara) [reply Runbooks/startstop/1.1/SequencedStartStop_Parent.ps1:10]: Fixed indentation

File: Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1 (hunk 15 above)
[the remark this answers is not in the document]
Comment (Rewdie) [reply Runbooks/startstop/1.0/AutoStop_CreateAlert_Child.ps1]: It looks ok now. Don't know what was up with the previous version.

**Review by Rewdie — APPROVED** [review]
Changes look good!

**Pull request merged**